recipe/install_nvcc.sh

```diff
@@ -2,61 +2,98 @@
-fi
-
 # Set `CUDA_HOME` in an activation script.
 mkdir -p "${PREFIX}/etc/conda/activate.d"
-cat > "${PREFIX}/etc/conda/activate.d/${PKG_NAME}_activate.sh" <<EOF
+cat >"${PREFIX}/etc/conda/activate.d/${PKG_NAME}_activate.sh" <<EOF
 #!/bin/bash
 
-if [[ ! -d "${CUDA_HOME}" ]]
+export CUDA_HOME_CONDA_NVCC_BACKUP="\${CUDA_HOME:-}"
+export CFLAGS_CONDA_NVCC_BACKUP="\${CFLAGS:-}"
+export CPPFLAGS_CONDA_NVCC_BACKUP="\${CPPFLAGS:-}"
+export CXXFLAGS_CONDA_NVCC_BACKUP="\${CXXFLAGS:-}"
+
+# Default to using \$(cuda-gdb) to specify \$(CUDA_HOME).
+if [ -z \${CUDA_HOME+x} ]
 then
-    echo "Cannot find: ${CUDA_HOME}"
-    exit 1
+    CUDA_GDB_EXECUTABLE=\$(which cuda-gdb)
+    if [ -n "\$CUDA_GDB_EXECUTABLE" ]; then
+        CUDA_HOME=\$(dirname \$(dirname \$CUDA_GDB_EXECUTABLE))
+    else
+        echo "Cannot determine CUDA_HOME: cuda-gdb not in PATH"
+        exit 1
+    fi
 fi
-if [[ ! -f "${CUDA_HOME}/lib64/stubs/libcuda.so" ]]
+
+if [[ ! -d "\${CUDA_HOME}" ]]
 then
-    echo "Cannot find: ${CUDA_HOME}/lib64/stubs/libcuda.so"
+    echo "Cannot find: \${CUDA_HOME}"
     exit 1
 fi
-grep "CUDA Version ${PKG_VERSION}" ${CUDA_HOME}/version.txt &>/dev/null
-if [[ $? -ne 0 ]]
+
+if [[ ! -f "\${CUDA_HOME}/lib64/stubs/libcuda.so" ]]
 then
+    echo "Cannot find: \${CUDA_HOME}/lib64/stubs/libcuda.so"
+    exit 1
+fi
+
+if [[ \$(grep -q "CUDA Version ${PKG_VERSION}" \${CUDA_HOME}/version.txt) -ne 0 ]]; then
     echo "Version of installed CUDA didn't match package"
     exit 1
 fi
 
-export CUDA_HOME="${CUDA_HOME}"
+export CUDA_HOME="\${CUDA_HOME}"
 export CFLAGS="\${CFLAGS} -I\${CUDA_HOME}/include"
 export CPPFLAGS="\${CPPFLAGS} -I\${CUDA_HOME}/include"
 export CXXFLAGS="\${CXXFLAGS} -I\${CUDA_HOME}/include"
+
+# Add \$(libcuda.so) shared object stub to the compiler sysroot.
+# Needed for things that want to link to \$(libcuda.so).
+# Stub is used to avoid getting driver code linked into binaries.
+
+CONDA_ENV_SYSROOT="\$(\${CC} --print-sysroot)"
+mkdir -p "\${CONDA_ENV_SYSROOT}/lib"
+ln -s "\${CUDA_HOME}/lib64/stubs/libcuda.so" "\${CONDA_ENV_SYSROOT}/lib/libcuda.so"
 EOF
 
 # Unset `CUDA_HOME` in a deactivation script.
 mkdir -p "${PREFIX}/etc/conda/deactivate.d"
-cat > "${PREFIX}/etc/conda/deactivate.d/${PKG_NAME}_deactivate.sh" <<EOF
+cat >"${PREFIX}/etc/conda/deactivate.d/${PKG_NAME}_deactivate.sh" <<EOF
 #!/bin/bash
-unset CUDA_HOME
-unset CFLAGS
-unset CPPFLAGS
-unset CXXFLAGS
+
+export CUDA_HOME="\${CUDA_HOME_CONDA_NVCC_BACKUP}"
+export CFLAGS="\${CFLAGS_CONDA_NVCC_BACKUP}"
+export CPPFLAGS="\${CPPFLAGS_CONDA_NVCC_BACKUP}"
+export CXXFLAGS="\${CXXFLAGS_CONDA_NVCC_BACKUP}"
+unset CUDA_HOME_CONDA_NVCC_BACKUP
+unset CFLAGS_CONDA_NVCC_BACKUP
+unset CPPFLAGS_CONDA_NVCC_BACKUP
+unset CXXFLAGS_CONDA_NVCC_BACKUP
+
+if [ -z "\${CUDA_HOME}" ]; then
+    unset CUDA_HOME
+fi
+
+if [ -z "\${CFLAGS}" ]; then
+    unset CFLAGS
+fi
+
+if [ -z "\${CPPFLAGS}" ]; then
+    unset CPPFLAGS
+fi
+
+if [ -z "\${CXXFLAGS}" ]; then
+    unset CXXFLAGS
+fi
```

Comment:
I don't think we want these given we have restored the backed up values above. Though I could be missing something.

Reply:
The whole point of having the `unset`-s here is to restore the original environment exactly.

(Although unlikely) when `CXXFLAGS` was not set in the parent environment before activation then without the `unset`, after deactivation `CXXFLAGS` will be defined in the parent environment (with no value), hence conda activate/deactivate is having a sideeffect.

Reply:
Makes sense. Thanks. Marking this as resolved.